File: drone_forest/drone.py
"""Script for the drone class."""

import cv2
import numpy as np

from drone_forest.geometric_objects import Point
from drone_forest.lidar import Lidar
from drone_forest.render_utils import transform_coords_to_render, IMAGE_HEIGHT


class Drone:
    """A drone object in the drone forest.

    Attributes:
        position (Point): The position of the drone.
        velocity (Point): The velocity of the drone.
        lidar (Lidar): The lidar attached to the drone.
    """

    def __init__(
        self,
        position: Point,
        lidar: Lidar,
        max_speed: float = 1.0,
        max_acceleration: float = 0.6,
    ):
        """Initialize the drone object.

        Args:
            position (Point): The position of the drone.
            lidar (Lidar): The lidar attached to the drone.
        """
        self.position: Point = position
        self.velocity: Point = Point(0, 0)
        self.max_speed: float = max_speed
        self.max_acceleration: float = max_acceleration
        self.lidar: Lidar = lidar
        self.update_lidar_position()

    def draw(self, img: np.ndarray, t_vec: Point, m2px: float):
        """Draw the drone on the given axis.

        Args:
            ax (matplotlib.axes.Axes): The axis to draw the drone on.
        """
        assert img is not None, "The axis must be provided."

        # Draw the lidar attached to the drone
        self.lidar.draw(img, t_vec, m2px)

        # Draw the drone as a small rectangle
        coord_bottom_left = transform_coords_to_render(
            Point(self.position.x - 0.05, self.position.y - 0.1), t_vec
        )
        coord_top_right = transform_coords_to_render(
            Point(self.position.x + 0.05, self.position.y + 0.1), t_vec
        )
        cv2.rectangle(
            img,
            (
                int(coord_bottom_left.x * m2px),
                int(coord_bottom_left.y * m2px) + IMAGE_HEIGHT,
            ),
            (
                int(coord_top_right.x * m2px),
                int(coord_top_right.y * m2px) + IMAGE_HEIGHT,
            ),
            (0, 0, 0),
            -1,
        )

    def move(self, dt: float, v: Point):
        """Move the drone by a given velocity.

        Args:
            dt (float): The time step for the movement.
            v (Point): The velocity of the drone.
        """
        acceleration = (v - self.velocity) / dt
        if acceleration.x > self.max_acceleration:
            acceleration.x = self.max_acceleration
        if acceleration.y > self.max_acceleration:
            acceleration.y = self.max_acceleration
        if acceleration.x < -self.max_acceleration:
            acceleration.x = -self.max_acceleration
        if acceleration.y < -self.max_acceleration:
            acceleration.y = -self.max_acceleration

        v = self.velocity + acceleration * dt
        if v.x > self.max_speed:
            v.x = self.max_speed
        if v.y > self.max_speed:
            v.y = self.max_speed
        if v.x < -self.max_speed:
            v.x = -self.max_speed
        if v.y < -self.max_speed:
            v.y = -self.max_speed

        self.velocity = v
        self.position += self.velocity * dt
        self.update_lidar_position()

    def update_lidar_position(self):
        """Update the position of the lidar attached to the drone."""
        self.lidar.position = self.position

File: src/drone.cc
#include <drone_forest/drone.h>

namespace evs
{
namespace drone
{

Drone::Drone(const geometric::Point& position, double lidar_range,
             int lidar_n_beams, double max_speed, double max_acceleration,
             double drone_width_m, double drone_height_m)
    : position_(position),
      velocity_(0.0, 0.0),
      max_speed_(max_speed),
      max_acceleration_(max_acceleration),
      lidar_(position, lidar_range, lidar_n_beams),
      body_(position + geometric::Point(drone_width_m / 2, drone_height_m / 2),
            position - geometric::Point(drone_width_m / 2, drone_height_m / 2))
{
}

geometric::Rectangle Drone::Body() const
{
  return body_;
}

void Drone::Draw(cv::Mat& image, geometric::Point t_vec, double m2px) const
{
  // Draw the LiDAR sensor
  lidar_.Draw(image, t_vec, m2px);

  // Draw the drone body (black rectangle)
  body_.Draw(image, cv::Scalar(0, 0, 0), t_vec, m2px);
}

lidar::Lidar Drone::Lidar() const
{
  return lidar_;
}

std::vector<double> Drone::LidarScan(
    const std::vector<geometric::Circle>& obstacles)
{
  return lidar_.Scan(obstacles);
}

geometric::Point Drone::Move(double dt, const geometric::Point& velocity)
{
  // Calculate the acceleration
  geometric::Point acceleration = (velocity - velocity_) / dt;
  if (acceleration.x() > max_acceleration_)
  {
    acceleration = geometric::Point(max_acceleration_, acceleration.y());
  }
  if (acceleration.y() > max_acceleration_)
  {
    acceleration = geometric::Point(acceleration.x(), max_acceleration_);
  }
  if (acceleration.x() < -max_acceleration_)
  {
    acceleration = geometric::Point(-max_acceleration_, acceleration.y());
  }
  if (acceleration.y() < -max_acceleration_)
  {
    acceleration = geometric::Point(acceleration.x(), -max_acceleration_);
  }

  // Calculate the new velocity
  velocity_ = velocity_ + acceleration * dt;
  if (velocity.x() > max_speed_)
  {
    velocity_ = geometric::Point(max_speed_, velocity_.y());
  }
  if (velocity.y() > max_speed_)
  {
    velocity_ = geometric::Point(velocity_.x(), max_speed_);
  }
  if (velocity.x() < -max_speed_)
  {
    velocity_ = geometric::Point(-max_speed_, velocity_.y());
  }
  if (velocity.y() < -max_speed_)
  {
    velocity_ = geometric::Point(velocity_.x(), -max_speed_);
  }

  // Calculate the new position
  geometric::Point position_delta = velocity_ * dt;
  geometric::Point new_position = position_ + position_delta;

  // Update the LiDAR sensor position
  lidar_.UpdatePosition(new_position);

  // Update the drone position
  body_.UpdatePosition(new_position);
  position_ = new_position;

  return position_;
}

geometric::Point Drone::Position() const
{
  return position_;
}

void Drone::Reset(const geometric::Point& position)
{
  // Reset the drone position
  position_ = position;
  body_.UpdatePosition(position);

  // Reset the drone velocity
  velocity_ = geometric::Point(0, 0);

  // Reset the LiDAR sensor
  lidar_.Reset(position_);
}

}  // namespace drone
}  // namespace evs